checkpoint/keeper_milestone.go

```diff
@@ -0,0 +1,208 @@
+		if err := k.cdc.UnmarshalBinaryBare(store.Get(milestoneKey), &_milestone); err != nil {
+			return nil, err
+		}
+
+		return &_milestone, nil
+	}
+
+	return nil, errors.New("Invalid milestone Index")
+}
+
+// GetLastMilestone gets last milestone, milestone number = GetCount()
+func (k *Keeper) GetLastMilestone(ctx sdk.Context) (*hmTypes.Milestone, error) {
+	store := ctx.KVStore(k.storeKey)
+	Count := k.GetMilestoneCount(ctx)
+
+	lastMilestoneKey := GetMilestoneKey(Count)
+
+	// fetch milestone and unmarshall
+	var _milestone hmTypes.Milestone
+
+	if store.Has(lastMilestoneKey) {
+		err := k.cdc.UnmarshalBinaryBare(store.Get(lastMilestoneKey), &_milestone)
+		if err != nil {
+			k.Logger(ctx).Error("Unable to fetch last milestone from store", "number", Count)
+			return nil, err
+		} else {
+			return &_milestone, nil
+		}
+	}
+
+	return nil, cmn.ErrNoMilestoneFound(k.Codespace())
+}
+
+// SetCount set the count number
+func (k *Keeper) SetMilestoneCount(ctx sdk.Context, number uint64) {
+	store := ctx.KVStore(k.storeKey)
+	// convert timestamp to bytes
+	value := []byte(strconv.FormatUint(number, 10))
+	// set no-ack
+	store.Set(CountKey, value)
+}
+
+// GetCount returns count
+func (k *Keeper) GetMilestoneCount(ctx sdk.Context) uint64 {
+	store := ctx.KVStore(k.storeKey)
+	// check if count is there
+	if store.Has(CountKey) {
+		// get current count
+		result, err := strconv.ParseUint(string(store.Get(CountKey)), 10, 64)
+		if err == nil {
+			return result
+		}
+	}
+
+	return uint64(0)
+}
+
+// FlushCheckpointBuffer flushes Checkpoint Buffer
+func (k *Keeper) PruneMilestone(ctx sdk.Context, number uint64) {
+	store := ctx.KVStore(k.storeKey)
+
+	if number <= 0 {
+		return
+	}
+
+	milestoneKey := GetMilestoneKey(number)
+
+	if !store.Has(milestoneKey) {
+		return
+	}
+
+	store.Delete(milestoneKey)
+}
+
+// SetLastNoAck set last no-ack object
+func (k *Keeper) SetNoAckMilestone(ctx sdk.Context, milestoneId string) {
+	store := ctx.KVStore(k.storeKey)
+
+	milestoneNoAckKey := GetMilestoneNoAckKey(milestoneId)
+	value := []byte(milestoneId)
+
+	// set no-ack-milestone
+	store.Set(milestoneNoAckKey, value)
+	store.Set(MilestoneLastNoAckKey, value)
+}
+
+// GetLastNoAckMilestone returns last no ack milestone
+func (k *Keeper) GetLastNoAckMilestone(ctx sdk.Context) string {
+	store := ctx.KVStore(k.storeKey)
+	// check if ack count is there
+	if store.Has(MilestoneLastNoAckKey) {
+		// get current ACK count
+		result := string(store.Get(MilestoneLastNoAckKey))
+		return result
+	}
+
+	return ""
+}
+
+// GetLastNoAckMilestone returns last no ack milestone
+func (k *Keeper) GetNoAckMilestone(ctx sdk.Context, milestoneId string) bool {
+	store := ctx.KVStore(k.storeKey)
+	// check if No Ack Milestone is there
+	if store.Has(GetMilestoneNoAckKey(milestoneId)) {
+		return true
+	}
+
+	return false
+}
+
+// SetLastMilestoneTimeout set lastMilestone timeout time
+func (k *Keeper) SetLastMilestoneTimeout(ctx sdk.Context, timestamp uint64) {
```

Comment:
Why we use uint as a timestamp, not time.Time?

Reply:
We need the time for just comparison purpose and uint is more optimized than time.Time in respect of storage and operations.

Reply:
Have you measuted it?